File: android/network/src/main/kotlin/com/ridelink/network/control/ReconnectPolicy.kt
package com.ridelink.network.control

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

/**
 * PROTOCOL §10: `0.5, 1, 2, 4, 8, 8, 8 … s` with ±20% jitter, budget 120 s total, then
 * `DISCONNECTED`. Pure — no clock reads, no [kotlinx.coroutines.delay] — so it is testable
 * without `Thread.sleep()` (CLAUDE.md rule 12 / this session's brief §12).
 */
object ReconnectPolicy {
    // PROTOCOL §10's ladder, spelled out literally — a formula would obscure the exact spec values.
    @Suppress("MagicNumber")
    private val LADDER_MS = longArrayOf(500, 1000, 2000, 4000, 8000)
    const val MAX_TOTAL_BUDGET_MS = 120_000L
    const val JITTER_FRACTION = 0.2

    /** `attempt` is 0-indexed; the ladder holds at its last (8 s) rung beyond its own length. */
    fun baseDelayMs(attempt: Int): Long = LADDER_MS.getOrElse(attempt) { LADDER_MS.last() }

    /** @param randomFraction in `[-1.0, 1.0]`, supplied by the caller's injected [kotlin.random.Random]. */
    fun jitteredDelayMs(
        attempt: Int,
        randomFraction: Double,
    ): Long {
        require(randomFraction in -1.0..1.0) { "randomFraction must be in [-1, 1]" }
        val base = baseDelayMs(attempt)
        val jitter = (base * JITTER_FRACTION * randomFraction).toLong()
        return (base + jitter).coerceAtLeast(0)
    }
}

/** Why a reconnect loop did or did not start (ARCHITECTURE §3 rule 6 / PROTOCOL §4.2, §4.6). */
enum class LinkLossReason { NETWORK, BYE, DUPLICATE_CONNECTION, USER_ENDED }

/**
 * Drives [ReconnectPolicy]'s ladder against real time via an injectable delay function, so unit
 * tests can replace [delayMs] with a no-op recorder instead of sleeping (CLAUDE.md rule 12).
 * `BYE`, `duplicate_connection` and a deliberate user end never reach this controller at all —
 * the caller (network) checks [LinkLossReason] before starting it, mirroring the FSM's own
 * suppression (`core.sessionfsm` `LinkLost(BYE)` / `DuplicateConnectionClosed`).
 */
class ReconnectController(
    private val scope: CoroutineScope,
    private val random: kotlin.random.Random,
    private val delayMs: suspend (Long) -> Unit,
) {
    private var job: Job? = null
    private var attempt = 0
    private var elapsedMs = 0L

    var reconnectCount: Int = 0
        private set

    /**
     * Starts (or restarts) the retry loop. [onAttempt] returns `true` on success (stops the
     * loop and resets it for next time) or `false` to keep retrying. [onExhausted] fires once
     * the 120 s budget is spent with no success.
     */
    fun start(
        onAttempt: suspend () -> Boolean,
        onExhausted: suspend () -> Unit,
    ) {
        cancel()
        job =
            scope.launch {
                while (elapsedMs < ReconnectPolicy.MAX_TOTAL_BUDGET_MS) {
                    val fraction = random.nextDouble(-1.0, 1.0)
                    val delay = ReconnectPolicy.jitteredDelayMs(attempt, fraction)
                    delayMs(delay)
                    elapsedMs += delay
                    attempt += 1
                    reconnectCount += 1
                    if (onAttempt()) {
                        reset()
                        return@launch
                    }
                }
                onExhausted()
            }
    }

    fun cancel() {
        job?.cancel()
        job = null
    }

    fun reset() {
        attempt = 0
        elapsedMs = 0
    }
}

File: ios/Packages/RideLinkPlatform/Sources/RideLinkPlatform/Control/ReconnectPolicy.swift
import Foundation

/// PROTOCOL §10: `0.5, 1, 2, 4, 8, 8, 8 … s` with ±20% jitter, budget 120s total, then
/// `DISCONNECTED`. Pure — no real clock reads, no `Task.sleep` — so it is testable without
/// sleeping (CLAUDE.md rule 12 / this session's brief §12).
public enum ReconnectPolicy {
    private static let ladderMs: [Int64] = [500, 1000, 2000, 4000, 8000]
    public static let maxTotalBudgetMs: Int64 = 120_000
    public static let jitterFraction = 0.2

    /// `attempt` is 0-indexed; the ladder holds at its last (8s) rung beyond its own length.
    public static func baseDelayMs(_ attempt: Int) -> Int64 {
        attempt < ladderMs.count ? ladderMs[attempt] : ladderMs[ladderMs.count - 1]
    }

    /// - Parameter randomFraction: in `[-1.0, 1.0]`, supplied by the caller's injected RNG.
    public static func jitteredDelayMs(_ attempt: Int, randomFraction: Double) -> Int64 {
        precondition(randomFraction >= -1.0 && randomFraction <= 1.0, "randomFraction must be in [-1, 1]")
        let base = baseDelayMs(attempt)
        let jitter = Int64(Double(base) * jitterFraction * randomFraction)
        return max(0, base + jitter)
    }
}

/// Why a reconnect loop did or did not start (ARCHITECTURE §3 rule 6 / PROTOCOL §4.2, §4.6).
public enum LinkLossReason: Sendable {
    case network
    case bye
    case duplicateConnection
    case userEnded
}

/// Drives `ReconnectPolicy`'s ladder against real time via an injectable delay function, so unit
/// tests can replace `delayNs` with a no-op recorder instead of sleeping (CLAUDE.md rule 12).
/// `BYE`, `duplicate_connection` and a deliberate user end never reach this controller at all —
/// the caller checks `LinkLossReason` before starting it.
public actor ReconnectController {
    private var attempt = 0
    private var elapsedMs: Int64 = 0
    private var task: Task<Void, Never>?

    public private(set) var reconnectCount = 0

    private let randomFraction: @Sendable () -> Double
    private let delayMs: @Sendable (Int64) async -> Void

    public init(randomFraction: @escaping @Sendable () -> Double, delayMs: @escaping @Sendable (Int64) async -> Void) {
        self.randomFraction = randomFraction
        self.delayMs = delayMs
    }

    /// Starts (or restarts) the retry loop. `onAttempt` returns `true` on success (stops the loop
    /// and resets it for next time) or `false` to keep retrying. `onExhausted` fires once the
    /// 120s budget is spent with no success.
    ///
    /// **Must check `Task.isCancelled` explicitly** (this session's brief §9/§10). `delayMs`'s
    /// production implementation is `try? await Task.sleep(...)` — the `try?` is there so a
    /// timer tick never throws into this loop, but it also swallows the `CancellationError`
    /// `Task.sleep` throws when `cancel()` cancels this task. Without an explicit check, a
    /// cancelled loop would keep running to completion in the background, invisible to
    /// `cancel()`'s caller and still calling `onAttempt` on a schedule no one asked for.
    public func start(onAttempt: @escaping @Sendable () async -> Bool, onExhausted: @escaping @Sendable () async -> Void) {
        cancel()
        task = Task {
            while self.remainingBudget() {
                if Task.isCancelled { return }
                let fraction = self.randomFraction()
                let (currentAttempt, delay) = self.nextDelay(fraction: fraction)
                await self.delayMs(delay)
                if Task.isCancelled { return }
                self.recordElapsed(delay, attempt: currentAttempt)
                if await onAttempt() {
                    self.reset()
                    return
                }
            }
            if !Task.isCancelled {
                await onExhausted()
            }
        }
    }

    private func remainingBudget() -> Bool { elapsedMs < ReconnectPolicy.maxTotalBudgetMs }

    private func nextDelay(fraction: Double) -> (Int, Int64) {
        let current = attempt
        return (current, ReconnectPolicy.jitteredDelayMs(current, randomFraction: fraction))
    }

    private func recordElapsed(_ delay: Int64, attempt currentAttempt: Int) {
        elapsedMs += delay
        attempt = currentAttempt + 1
        reconnectCount += 1
    }

    public func cancel() {
        task?.cancel()
        task = nil
    }

    public func reset() {
        attempt = 0
        elapsedMs = 0
    }
}
